### DISCUSSION TITLE
grad_test with Tri-mesh and RZ coords

### DISCUSSION THREAD

**Discussion by joe61vette** [idaholab/moose#23929] [Q&A Modules: General] [answered]
Hello:

I have an app where velocity is "constant monomial" and pressure is "Lagrange first".  This is a hybrid rDG/CG scheme that works great most of the time.  But...

computeQpResidual() for the mass conservation equation (inherited from ADKernel) whose non-linear variable is pressure has:

res = _rho[_qp] * _velocity[_qp] * _grad_test[_i][_qp];

where velocity is an ADMaterialProperty(RealVectorValue) whose components are vel_x and vel_y for this 2D problem.  The problem is a simple rectangle with uniform inlet velocity and outlet pressure BCs. I have simplified the model down so there is no momentum advection, no gravity.  Just pressure gradient and form drag. 

It works great when:

1) The mesh is all quads and the coordinate system is either XY or RZ.  No problem.
2) The mesh is all tris and the coord system is XY.

It does not work if the coord system is RZ and the mesh includes tris.  To make sure this wasn't an inlet BC problem, I put a quad mesh as an inlet section and have it transition to a tri-mesh.  The attached 2 figures show the velocity magnitude for both XY and RZ systems.  The velocity should be uniform (all ve_y) with a value of 0.1.  Note the higher velocity magnitude for the RZ case for the first tri at the centerline that is due to a vel_x component.  It appears that this is caused by the residual of this mass eq for the first tri element.  It should be zero but instead has a small value that forces the radial (vel_x) velocity to be non-zero.  The flow does go to a uniform distribution further downstream.

Finally, if I offset these blocks from the centerline in the RZ system, the unphysical vel_x component gradually disappears as the block is moved further away from the centerline.

Any ideas?  Help is greatly appreciated.
  Joe Kelly

![XY-coords](https://user-images.githubusercontent.com/8108870/229293105-42dc3bb4-88bc-4e1f-97c8-1d0f1f4ca368.png)

![RZ-coords](https://user-images.githubusercontent.com/8108870/229293110-cb48b73a-10cd-418a-ab80-d32871765e5b.png)

**Comment by GiudGiud** [comment, marked as answer]
Hello

In RZ, there should be an extra term for the pressure gradient, just arising from the definition of the divergence in RZ

For the first component of momentum RZ, if R is the first components (usually like this)
res += -_pressure[_qp] / _ad_q_point[_qp](0) * _test[_i][_qp];

Do you have it here?

Guillaume

**Reply by GiudGiud** [reply]
For XY with a mix of quads and tris, this should have worked, we would definitely need to investigate that.
Have you tried SVD, do you know how many singular values you have? This helps guessing where the issue happens

**Reply by joe61vette** [reply]
Hi Guillaume:  For the pressure gradient term, I am not integrating by parts, so I am using:

return _eps[_qp] * _grad_pressure[_qp](_component) * _test[_i][_qp];

where the non-linear vars are vel_x and vel_y.  I looked at this first, but it works perfectly for quads in RZ.  I'll try svd and get back to you.  Just about done for my Saturday workday...

  Joe

**Reply by joe61vette** [reply]
For this problem, I have eliminated the momentum flux terms (for debug purposes).  So, in effect, I have turned this into a Poisson problem.  With constant fluid properties, the mass eq. simply states that div(V) = 0.  The momentum eq, is simply grad(P) = -K|V|V.

At any rate, here is the SVD output for 1st iteration:

     SVD: condition number 6.229482285046e+03, 0 of 56 singular values are (nearly) zero
      SVD: smallest singular values: 1.673742906471e-04 7.262316760914e-04 1.217246675124e-03 1.653500109633e-03 2.200366856638e-03
      SVD: largest singular values : 1.000004573129e+00 1.000082403684e+00 1.000240952341e+00 1.041314956785e+00 1.042655178559e+00

I can provide the Jacobin as well.  But everything works perfectly in RZ as long as I don't use Tri's.

 Joe

**Reply by joe61vette** [reply]
Is there a good reason to integrate by parts for the pressure gradient term?  I would prefer not to as I would also need the porosity gradient.  That would cause the porosity to be an auxvar instead of a material property.

As I do some more complex geometries, even with a quad mesh, albeit one that is can be slightly skewed, I begin to see a bit of striping in the pressure in RZ.  No striping in XY.

**Reply by GiudGiud** [reply]
Ok in the mass equation you did integrate by parts right?
`res = _rho[_qp] * _velocity[_qp] * _grad_test[_i][_qp];`
this is how you got v out of the divergence and grad_test 

~~It should have looked like integral[ 1/r d(r rho v)/dr * test * rdr ] = integral[ d(r rho v)/dr * test dr] -> int [ (r rho v) * grad_test] + boundary term~~

~~so is there a missing r factor?~~
Nevermind that factor comes later, multiplying the contribution to the residual

In INS and INSAD we dont integrate it by parts.
Do you compare well to INS and INSAD results for simple RZ channels?

I'm not sure why those terms are integrated by parts or not. For diffusion terms, it's to reduce the FE order required by one. 
Here it may just be to make boundary terms more or less simple.

**Reply by GiudGiud** [reply]
Made a mesh to investigate that.
Works fine with INS with second order velocities, first order pressure
```
[Mesh]
  [inlet]
    type = CartesianMeshGenerator
    ix = 2
    iy = 4
    dim = 2
    dx = 1
    dy = 1
    subdomain_id = 2
  []

  ## add some tris
  [poly]
    type = PolyLineMeshGenerator
    points = '0.0 1.0 0.0
              0.0 2.0 0.0
              1.0 2.0 0.0
              1.0 1.0 0.0'
    loop = true
    num_edges_between_points = 2
  []
  [triang]
    type = XYDelaunayGenerator
    boundary = 'poly'
    refine_boundary = false
    desired_area = 0.2
  []
  [bot]
    type = SideSetsFromNormalsGenerator
    input = triang
    normals = '0 -1 0
               1 0 0
               -1 0 0
               0 1 0'
    new_boundary = 'bottom_tri
                    left
                    right
                    top_tri'
    replace = true
  []

  # Stitch
  [stitch]
    type = StitchedMeshGenerator
    inputs = 'inlet bot'
    stitch_boundaries_pairs = 'top bottom_tri'
    clear_stitched_boundary_ids = true
  []
  [rename_error]
    type = RenameBoundaryGenerator
    input = 'stitch'
    old_boundary = 'top'
    new_boundary = 'right'
  []
[]
```

**Reply by joe61vette** [reply]
Thanks for taking the time to investigate this Guillaume.  I am pretty convinced that the usual weak form of the mass advection term, namely:

res = _rho[_qp] * _velocity[_qp] * _grad_test[_i][_qp];

isn't quite perfect for RZ geometry when the velocity vars are "constant monomials" and hence essentially cell-centered finite volume vars.

I looked at INSADMass & INSMassRZ etc.  So, if one uses the strong form, then it is necessary to add the U/r term.  I thought about giving this a try but I can't use the gradient of the velocity as that is zero for monomials.

What I would really like to be able to do is use the ADDGKernel to compute the mass advection term consistent with how it is done for momentum and energy.  But my non-linear var for the mass eq. is pressure which is "Gauss first" and hence FEM.  So, I need to use CG type kernels for that.

There is probably some way to accomplish that, using DG type advection with CG, but it will take me into parts of Moose that I don't have an understanding for.  I might need to do that though when I have a good bit more time.

Thanks again for the help,
  Joe

**Reply by joe61vette** [reply]
It does appear that I have a problem with skewed mesh cells, even with X-Y coords.  It really shows up when the energy eq. is turned on.  I am guessing that it is necessary for the mass/momentum/energy advection terms to all have consistent fluxes.  That is, use a DGKernel flux across the faces for all eqs.  

So, the question of whether or not I can have P as a CFEM variable with others as DG (FV) is likely the real problem.  With mesh cells that are perpendicular to X-Y, the current scheme works well and avoids the checkerboard problem w/o resorting to a something like RC.

Looks like I have some work ahead of me to make this work.

Thanks again for the help.   Joe

**Reply by GiudGiud** [reply]
@lindsayad thoughts?

> There is probably some way to accomplish that, using DG type advection with CG

I think I agree that we d want the conservation of mass equation to look much like the conservation of momentum and energy, with the same kind of fluxes, differing only by the appropriate factor.

To do that we would need the appropriate set of test functions for that equation, maybe the same as for velocity? This seems to me like we would want to use different FE families for solution (pressure) and test functions (to write the fluxes), so a Petrov Galerkin method.

**Reply by GiudGiud** [reply]
I wonder if, as an exploration, we could have the following.
- add dummy nonlinear variable Z that is of the same type as velocity
- use Z for the variable parameter of the mass equation, write the mass equation in a DG form
- use pressure for the variable parameter of the Z equation, which is just dZ/dt = 0
This will off-course have a lot of off-diagonal terms, but a direct solver (LU with a diagonal shift for now) would handle them

**Reply by lindsayad** [reply]
A few questions:

- Does all the literature use first order Lagrange for pressure and constant-monomial/piecewise-constant (with reconstruction) for the velocity? If so I find that interesting given that there will be more degrees of freedom for the Lagrange multiplier constraint variable (pressure), than for the primal variable (velocity).
- You're not doing anything special for the evaluation of `_velocity[_qp]` in the mass kernel are you? If not, then we're just evaluating a constant at different quadrature points. I would expect that to reduce the order of accuracy somewhere.

I agree that if you are using constant monomials with reconstruction for the velocity that you are going to have the same issue with skew that we have for finite volume (indeed the discretization is the exact same). In that vein, have you considered using first order monomial DG instead of rDG? Then you would have more accurate face values without the need for an explicit skew correction. Moreover, you could then work entirely in a finite element paradigm.

If you want to continue with rDG and work with skewed meshes, then I'd think hard about whether you (with our support) could reuse all that we've done with finite volume for discretizing the velocity equation. This gives you our skew correction. The only thing that would need to be modified in the velocity discretization is the pressure term, and even then you may not have to given that we perform a functor evaluation for the pressure gradient. We would just have to make sure we implement `evalauteGradient` with an elemental argument for finite element variables.

**Reply by joe61vette** [reply]
Hi Alex:

I am making this problem as simple as possible.  The idea is to get something that works and then refine it.  So, no reconstruction at all, just simple cell-centered constant monomials for velocity and temperature and upwind for the numerical flux.  To avoid the pressure checkerboard problem, from the Pandare reference you sent me, the pressure variable is first Lagrange.  This gives a value for the pressure gradient at the cell-center.

This scheme works GREAT for meshes that align with X-Y.  Note that I am not including shear stresses, this is more a porous media model where I have a drag force due to channel walls.  it converges very well and verification tests look great.  But..

When I started running more complex geometries, I began to see some problems.  I originally thought it had something to do with R-Z.  But it turned out to be a problem with skewed meshes.  With Guillaume's help. I was (eventually) able to figure out that the problem was due to an inconsistency between a velocity field that would conserve mass in the CG context vs. a velocity field that would conserve mass in a DG context.  For example, see the attached figure.  It is a fluid temperature taken after only 5 time steps for a constant density ISOTHERMAL problem.  

![T_fluid](https://user-images.githubusercontent.com/8108870/229819361-3e0b1cde-ac31-4a93-88b0-df3b3972e453.png)

There are no heat sources.  So, why does the one cell heat up?  The residual of the mass equation (solved for pressure using CG) is 10^-12.  But, if I hand calculate what the mass flows would be for a DG scheme (basically V dot n for the cell faces) I can see that there is 10% more mass leaving the cell than coming in.  So, the fluid temperature takes off immediately.  This also affects momentum transfer but as this is a relatively low velocity problem with high drag, that error is not so obvious.

To make this work, what I need is for my mass equation (pressure, 1st Lagrange) to be able to use the residual due to mass advection that is calculated by an ADDGKernel.  So, that is what I'll have to figure out.

From the other very good results I got, I believe this scheme holds promise for my two-phase app so that I don't have to implement a two-phase RC interpolation object.  I am working with single-phase now just to make sure it works.

  Joe

**Reply by joe61vette** [reply]
Note that the above is a long winded way of saying that the (divergence of density times the velocity vector) does not equal the (sum of the upwinded mass fluxes).

**Reply by lindsayad** [reply]
> To make this work, what I need is for my mass equation (pressure, 1st Lagrange) to be able to use the residual due to mass advection that is calculated by an ADDGKernel. So, that is what I'll have to figure out.

I would just make sure that you're planning to do matches literature practices. What you're proposing sounds pretty wonky to me; it doesn't sound variationally consistent but maybe I'm wrong.

You didn't answer my previous question about how you're evaluating the velocity in the mass equation. If you're just evaluating the constant monomial for velocity at every quadrature point, then yea I'm not surprised at all that the mass equation in a CG discretization is leading to really bad mass conservation in a DG context.

**Reply by lindsayad** [reply]
This is really interesting work that you're doing @joe61vette, and frankly I'm a bit jealous 😄 So I want to make sure we can get this right together
